File: dotnet/DesignPatterns/BehavioralDesignPatterns/CommandDesignPattern/CommandDesignPattern.Example1/CommandManager.cs
namespace CommandDesignPattern.Example1;

/// <summary>
/// Invoker - Manages command execution and undo/redo history
/// </summary>
public class CommandManager
{
    private readonly Stack<ICommand> _undoStack = new();
    private readonly Stack<ICommand> _redoStack = new();

    public void ExecuteCommand(ICommand command)
    {
        command.Execute();
        _undoStack.Push(command);
        _redoStack.Clear(); // Clear redo stack when a new command is executed
    }

    public void Undo()
    {
        if (_undoStack.Count == 0)
        {
            Console.WriteLine("Nothing to undo!");
            return;
        }

        var command = _undoStack.Pop();
        command.Unexecute();
        _redoStack.Push(command);
        Console.WriteLine("Undo completed.");
    }

    public void Redo()
    {
        if (_redoStack.Count == 0)
        {
            Console.WriteLine("Nothing to redo!");
            return;
        }

        var command = _redoStack.Pop();
        command.Execute();
        _undoStack.Push(command);
        Console.WriteLine("Redo completed.");
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/CommandDesignPattern/CommandDesignPattern.Example1/src/CommandManager.ts
import { ICommand } from "./ICommand";

/**
 * Invoker — manages command execution and undo/redo history.
 */
export class CommandManager {
  private readonly _undoStack: ICommand[] = [];
  private readonly _redoStack: ICommand[] = [];

  executeCommand(command: ICommand): void {
    command.execute();
    this._undoStack.push(command);
    this._redoStack.length = 0; // clear redo stack on new command
  }

  undo(): void {
    if (this._undoStack.length === 0) {
      console.log("Nothing to undo!");
      return;
    }
    const command = this._undoStack.pop()!;
    command.unexecute();
    this._redoStack.push(command);
    console.log("Undo completed.");
  }

  redo(): void {
    if (this._redoStack.length === 0) {
      console.log("Nothing to redo!");
      return;
    }
    const command = this._redoStack.pop()!;
    command.execute();
    this._undoStack.push(command);
    console.log("Redo completed.");
  }
}
